File: training/train on 1600-1800 problems/B_Array_Walk.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

int n,k,z;
ll dp[100005][6][2];
vector<int> a;

ll solve(int r,int l,int s){
    if(r+l==k) return 0;
    if (dp[r][l][s]!=-1) return dp[r][l][s];
    ll ch1=-1e9,ch2=-1e9;
    if (r-l+1<n) ch1=a[r-l+1]+solve(r+1,l,0);
    if (r-l-1>=0 && s!=1 && l<z) ch2=a[r-l-1]+solve(r,l+1,1);
    ll res=max(ch1,ch2);
    return dp[r][l][s]=res;
}

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    int t;
    cin>>t;
    while (t--){
        cin>>n>>k>>z;
        a.resize(n);
        for(int i=0;i<n;i++){
            cin>>a[i];
        }
        for (int i=0;i<=k;i++) {
            for (int j=0;j<=z;j++) {
                dp[i][j][0]=-1;
                dp[i][j][1]=-1;
            }
        }
        cout<<a[0]+solve(0,0,0)<<"\n";
    }
    return 0;
}

File: training/train on 1600-1800 problems/B_Array_Walk.py
import sys 
sys.setrecursionlimit(100005)

def solve(r,l,s):
    if r+l==k:
        return 0
    if dp[r][l][s]!=-1:
        return dp[r][l][s]
    ch1=-float("inf")
    ch2=-float("inf")
    if r-l+1<n:
        ch1=a[r-l+1]+solve(r+1,l,0)
    if r-l-1>=0 and s!=1 and l<z:
        ch2=a[r-l-1]+solve(r,l+1,1)
    res=max(ch1,ch2)
    dp[r][l][s]=res
    return res

t=int(input())
for _ in range(t):
    n,k,z=map(int,input().split())
    a=list(map(int,input().split()))
    dp=[[[-1]*2 for _ in range(z+1)] for _ in range(k+1)]
    print(a[0]+solve(0,0,0))